docs/content/concepts/resources.mdx

```diff
@@ -5,357 +5,499 @@ description: Resources enable you to separate graph logic from environment, and
 
-@resource(required_resource_keys={"foo"}, config_schema={"bar": str})
-def my_resource_requires_context(init_context):
-    return init_context.resources.foo, init_context.resource_config["bar"]
+### Adapting function-style resources
 
+In codebases that utilize [function-style](https://docs.dagster.io/concepts/resources#resources) resources with the <PyObject object="@resource"/> decorator, you can use the <PyObject object="ConfigurableResourceAdapter"/> class to adapt these resources to the new Pythonic resource pattern.
 
-def test_my_resource_with_context():
-    init_context = build_init_resource_context(
-        resources={"foo": "foo_str"}, config={"bar": "bar_str"}
-    )
-    assert my_resource_requires_context(init_context) == ("foo_str", "bar_str")
-```
+The adapter subclass should list all config fields as attributes and implement a `wrapped_resource` property which returns the underlying resource definition to adapt.
 
-If your resource is a context manager, then you can open it as one using python's `with` syntax.
+```python file=/concepts/resources/pythonic_resources.py startafter=start_resource_adapter endbefore=end_resource_adapter dedent=4
+from dagster import (
+    resource,
+    Definitions,
+    ResourceDefinition,
+    asset,
+    ConfigurableLegacyResourceAdapter,
+)
 
-```python file=/concepts/resources/resources.py startafter=start_test_cm_resource endbefore=end_test_cm_resource
-from contextlib import contextmanager
-from dagster import resource
+# Old code, interface cannot be changed for back-compat purposes
+class Writer:
+    def __init__(self, prefix: str):
+        self._prefix = prefix
 
+    def output(self, text: str) -> None:
+        print(self._prefix + text)
 
-@resource
-@contextmanager
-def my_cm_resource(_):
-    yield "foo"
+@resource(config_schema={"prefix": str})
+def writer_resource(context):
+    prefix = context.resource_config["prefix"]
+    return Writer(prefix)
 
+# New adapter layer
+class WriterResource(ConfigurableLegacyResourceAdapter):
+    prefix: str
 
-def test_cm_resource():
-    with my_cm_resource(None) as initialized_resource:
-        assert initialized_resource == "foo"
-```
+    @property
+    def wrapped_resource(self) -> ResourceDefinition:
+        return writer_resource
 
----
+@asset
+def my_asset(writer: Writer):
+    writer.output("hello, world!")
 
-## Initializing resources outside of execution
+defs = Definitions(
+    assets=[my_asset], resources={"writer": WriterResource(prefix="greeting: ")}
+)
+```
 
-There are scenarios where you might want to reuse the code written within your resources outside of the context of execution. Consider a case where you have a resource `db_connection`, and you want to use that resource outside of the context of an execution. You can use the <PyObject object="build_resources"/> API to initialize this resource outside of execution.
+### Adapting function-style I/O managers
```

Comment:
Should the `Defining Pythonic I/O managers` section link to this section? Something like "If you're utilizing function-style I/O managers, [click here for more info]"?

Reply:
Good idea, added!